plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,611 @@
+  auto I_eff = input_size;
+  auto F_eff = (dilation * (filter_size - 1)) + 1;
+  if (autopad_mode == plaidml::op::AutoPadMode::EXPLICIT) {
+    size_t output_size = ((I_eff + pad_lo + pad_hi - F_eff + stride) / stride);
+    return output_size;
+  }
+  if (autopad_mode == plaidml::op::AutoPadMode::VALID) {
+    size_t output_size = ((I_eff - F_eff + stride) / stride);
+    return output_size;
+  }
+  if (autopad_mode == plaidml::op::AutoPadMode::SAME_LOWER || autopad_mode == plaidml::op::AutoPadMode::SAME_UPPER) {
+    size_t output_size = ((I_eff + stride - 1) / stride);
+    return output_size;
+  }
+  THROW_IE_EXCEPTION << "Unexpected autopadding mode.";
+}
+// Extract values which is needed.
+edsl::Tensor extract_tensor(edsl::Tensor I, size_t rank) {
+  std::vector<TensorDim> I_dims(rank * (rank + 2) + 2), O_dims(rank + 2);
+  std::vector<TensorIndex> I_idxs(rank * (rank + 2) + 2), O_idxs(rank + 2);
+  I.bind_dims(I_dims);
+  // Set up the dimensions of output.
+  for (auto i = 2; i < rank + 4; ++i) {
+    O_dims[i - 2] = I_dims[i];
+  }
+  // Set up the indexs of batch_size and channel.
+  for (auto i = 0; i < 2; ++i) {
+    I_idxs[i + 2] = I_idxs[i];
+  }
+  // Set up the indexs of input.
+  for (auto i = 2; i < rank + 4; ++i) {
+    for (auto j = 1; j < rank; ++j) {
+      I_idxs[j * (rank + 2) + i] = I_idxs[i];
+    }
+  }
+  // Set up the indexs of output.
+  for (auto i = 2; i < rank + 4; ++i) {
+    O_idxs[i - 2] = I_idxs[i];
+  }
+  edsl::Tensor O = Contraction(O_dims, O_idxs).assign(I(I_idxs));
+  return O;
+}
+
+void registerDeformableConvolution() {
+  registerOp("DeformableConvolution", [](const Context& ctx) {
+    auto* layer = ngraph::as_type<ngraph::opset4::DeformableConvolution>(ctx.layer);
+    DType precision = to_plaidml(layer->get_output_element_type(0));
+    IE_ASSERT(ctx.operands.size() == 3);
+    auto I = ctx.operands.at(0), OFF = ctx.operands.at(1), F = ctx.operands.at(2);
+    auto I_shape = I.compute_shape().sizes(), OFF_shape = OFF.compute_shape().sizes(),
+         F_shape = F.compute_shape().sizes();
+    auto N = I_shape[0], CI = I_shape[1], CO = F_shape[0], OFF_C = OFF_shape[1];
+    auto G = layer->get_group(), DG = layer->get_deformable_group();
+    auto rank = I.rank();
+    // Get autopad_mode.
+    auto autopad_mode = to_plaidml(layer->get_auto_pad());
+    // Compute manual_padding.
+    std::vector<size_t> manual_padding;
+    if (autopad_mode == plaidml::op::AutoPadMode::EXPLICIT) {
+      for (auto pad : layer->get_pads_begin()) {
+        manual_padding.push_back(pad);
+      }
+      for (auto pad : layer->get_pads_end()) {
+        manual_padding.push_back(pad);
+      }
+      while (manual_padding.size() < 2 * (rank - 2)) {
+        manual_padding.push_back(0);
+      }
+    } else {
+      while (manual_padding.size() < 2 * (rank - 2)) {
+        manual_padding.push_back(0);
+      }
+    }
+    // Get the strides.
+    std::vector<size_t> strides;
```

Comment:
Can you use `auto strides=layer->get_strides();` ?

Reply:
ok, I will fix it.